File: server/apps/nexotype/schemas/knowledge_graph/genomic_association_schemas.py
from pydantic import BaseModel, ConfigDict
from datetime import datetime

# ==========================================
# Request Schemas
# ==========================================

class GenomicAssociationCreate(BaseModel):
    """Schema for creating a new genomic association"""
    variant_id: int
    indication_id: int
    odds_ratio: float | None = None


class GenomicAssociationUpdate(BaseModel):
    """Schema for updating a genomic association"""
    variant_id: int = None
    indication_id: int = None
    odds_ratio: float | None = None


# ==========================================
# Response Schemas
# ==========================================

class GenomicAssociationDetail(BaseModel):
    """Schema for genomic association details"""
    id: int
    variant_id: int
    indication_id: int
    odds_ratio: float | None = None
    created_at: datetime
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class GenomicAssociationListResponse(BaseModel):
    """Response schema for listing genomic associations"""
    success: bool
    data: list[GenomicAssociationDetail] | None = None
    count: int = 0
    error: str | None = None


class GenomicAssociationResponse(BaseModel):
    """Response schema for single genomic association operations"""
    success: bool
    data: GenomicAssociationDetail | None = None
    error: str | None = None


class MessageResponse(BaseModel):
    """Simple message response"""
    success: bool
    message: str | None = None
    error: str | None = None

File: client/src/modules/nexotype/schemas/knowledge_graph/genomic-association.schemas.ts
import { z } from 'zod';

// ==========================================
// Full Schema (GET responses)
// ==========================================

export const GenomicAssociationSchema = z.object({
  id: z.number(),
  variant_id: z.number().int(),
  indication_id: z.number().int(),
  odds_ratio: z.number().nullable().optional(),
  created_at: z.string(),
  updated_at: z.string().nullable(),
});

// ==========================================
// Input Schemas (POST/PUT)
// ==========================================

export const CreateGenomicAssociationSchema = z.object({
  variant_id: z.number().int(),
  indication_id: z.number().int(),
  odds_ratio: z.number().nullable().optional(),
});

export const UpdateGenomicAssociationSchema = z.object({
  variant_id: z.number().int().optional(),
  indication_id: z.number().int().optional(),
  odds_ratio: z.number().nullable().optional(),
});

// ==========================================
// Types
// ==========================================

export type GenomicAssociation = z.infer<typeof GenomicAssociationSchema>;
export type CreateGenomicAssociation = z.infer<typeof CreateGenomicAssociationSchema>;
export type UpdateGenomicAssociation = z.infer<typeof UpdateGenomicAssociationSchema>;

// ==========================================
// Response Types
// ==========================================

export type GenomicAssociationResponse = {
  success: boolean;
  data?: GenomicAssociation;
  error?: string;
};

export type GenomicAssociationsResponse = {
  success: boolean;
  data?: GenomicAssociation[];
  count?: number;
  error?: string;
};
